src/main/java/org/miracum/etl/fhirtoomop/mapper/ObservationMapper.java

```diff
@@ -1754,22 +1765,32 @@ private void setValueQuantityInMeasurement(
     basisMeasurement.setValueAsNumber(valueQuantity.getValue());
     basisMeasurement.setValueSourceValue(valueQuantity.getValue().toString());
 
-    var quantityUnitCodingFormat =
-        new Coding().setCode(valueQuantity.getCode()).setSystem(valueQuantity.getSystem());
-
-    var valueQuantityUnitConcept =
-        findOmopConcepts.getConcepts(
-            quantityUnitCodingFormat,
-            effectiveDateTime.toLocalDate(),
-            bulkload,
-            dbMappings,
-            observationId);
-    if (valueQuantityUnitConcept != null) {
-      basisMeasurement.setUnitConceptId(valueQuantityUnitConcept.getConceptId());
-      basisMeasurement.setUnitSourceValue(
-          valueQuantity.getUnit() == null
-              ? valueQuantityUnitConcept.getConceptCode()
-              : valueQuantity.getUnit());
+    if(valueQuantity.hasCode() && valueQuantity.hasSystem()){
+      var quantityUnitCodingFormat =
+              new Coding().setCode(valueQuantity.getCode()).setSystem(valueQuantity.getSystem());
+      var valueQuantityUnitConcept =
+              findOmopConcepts.getConcepts(
+                      quantityUnitCodingFormat,
+                      effectiveDateTime.toLocalDate(),
+                      bulkload,
+                      dbMappings,
+                      observationId);
+      if (valueQuantityUnitConcept != null) {
+
+        basisMeasurement.setUnitConceptId(valueQuantityUnitConcept.getConceptId());
+        basisMeasurement.setUnitSourceValue(
+                valueQuantity.getUnit() == null
+                        ? valueQuantityUnitConcept.getConceptCode()
+                        : valueQuantity.getUnit());
+        addToList(measurements, basisMeasurement);
+      }
+    }else{
+      if(valueQuantity.hasCode()){
+        basisMeasurement.setUnitSourceValue(valueQuantity.getCode());
+      }
+      if(valueQuantity.hasUnit()){
+        basisMeasurement.setUnitSourceValue(valueQuantity.getUnit());
+      }
```

Comment:
The same code is written to set the valueQuantity in Observation above. Can you extract it to a function?